ValueObject/ValueObject.cs

```diff
@@ -55,24 +53,26 @@ private bool FieldsAreEqual(object obj, FieldInfo f)
 
         private IEnumerable<PropertyInfo> GetProperties()
         {
-            if (this.properties == null)
+            Type type = GetType();
+            if (!properties.ContainsKey(type))
             {
-                this.properties = GetType().GetProperties(BindingFlags.Instance | BindingFlags.Public)
+                properties[type] = GetType().GetProperties(BindingFlags.Instance | BindingFlags.Public)

```

Comment:
since Dictionary is not thread safe, multiple threads can insert the same type here and break the dictionary